Refactor generateStatus, generateSize and generateDelay

thekuwayama opened the pull request:
Resolves #145 

This PR adds function to extract common code of generateStatus, generateSize and generateDelay.

thekuwayama committed:
Refactor generateStatus, generateSize and generateDelay using parseFromFormattedString

thekuwayama committed:
Add parseEntry && rename parseFormattedString

thekuwayama commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,106 @@ func removeTrailingPercent(s string) string {
-func generateStatus(status string) int {
-	lst := strings.Split(status, ",")
-	log.Debugf("Parsing status %s -> %v", status, lst)
+func parseFromFormattedString(input, about string, parseFunc func(string) (int64, error)) (int64, error) {
```
👍

thekuwayama commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,106 @@ func removeTrailingPercent(s string) string {
-func generateStatus(status string) int {
-	lst := strings.Split(status, ",")
-	log.Debugf("Parsing status %s -> %v", status, lst)
+func parseFromFormattedString(input, about string, parseFunc func(string) (int64, error)) (int64, error) {
```
resolved 👍

thekuwayama commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,118 @@ func removeTrailingPercent(s string) string {
+// input="XX:20,YY:10,ZZ:0.5" for 20% parsed_XX, 10% parsed_YY, 0.5% parsed_ZZ 69.5% -1(means default value)
+// entry parsed by parseFunc
+// when parseFromFormattedString raise error, return -1, error(not nil)
+func parseFormattedString(input, about string, parseFunc func(string) (int64, error)) (int64, error) {
```
resolved 👍

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,132 @@ func removeTrailingPercent(s string) string {
+// input="XX:20,YY:10,ZZ:0.5" for 20% parsed_XX, 10% parsed_YY, 0.5% parsed_ZZ 69.5% -1(means default value)
+// entry parsed by parseFunc
+// when parseFromFormattedString raise error, return -1, error(not nil)
+func parseFormattedString(input, about string, parseFunc func(string) (int64, error)) (int64, error) {
```
ignore this, it's fine

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -232,48 +232,60 @@ func parseEntry(entry, about, input string, parseFunc func(input string) (int64,
+		s, p, e := parseEntry(entry, about, input, parseFunc)
+		if e != nil {
+			err = e
+			goto ParseError
```
I'm not a big fan of goto but maybe in this case it's ok (usually return/break is enough- sometime adding an intermediate function helps too)

thekuwayama replied:
@ldemailly Thanks for the feedback !!
I rewrote.
Please let me know if you have any concerns.

thekuwayama committed:
Add getWeightedValue

thekuwayama commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,132 @@ func removeTrailingPercent(s string) string {
+// input="XX:20,YY:10,ZZ:0.5" for 20% parsed_XX, 10% parsed_YY, 0.5% parsed_ZZ 69.5% -1(means default value)
+// entry parsed by parseFunc
+// when parseFromFormattedString raise error, return -1, error(not nil)
+func parseFormattedString(input, about string, parseFunc func(string) (int64, error)) (int64, error) {
```
resolved 👍

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
 		if err != nil {
-			log.Warnf("Bad input size %v -> %v, not a number before colon", sizeInput, l2[0])
-			return size
+			log.Warnf("Bad input status %v -> %v, not a number before colon", input, s)
```
should be size, not status here

thekuwayama replied:
Thank you!
Done.

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
+	res := 100. * rand.Float32()
+	for i, v := range weights {
+		if res <= v {
+			if i >= len(values) {
```
this test/if is new, why / when does this case happen ? (comment needed at least)
you could refactor to have one vector of pairs / struct instead of 2 arrays - or assert the 2 arrays have the same size instead of silently swallowing what must be a bug of usage)

thekuwayama replied:
added WeightValuePair struct instead of 2 arrays.

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
+	if len(lst) == 1 && !strings.ContainsRune(input, ':') {
+		s, err := parseFunc(input)
 		if err != nil {
-			log.Warnf("Bad input status %v, not a number nor comma and colon separated %% list", status)
```
that warning is gone/lost ? (the case about being just a value that doesn't parse)

thekuwayama replied:
sorry, my mistake. add warning.

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
 		if err != nil {
-			log.Warnf("Bad input status %v, not a number nor comma and colon separated %% list", status)
-			return http.StatusBadRequest
+			s = -1
```
in this generic function you shouldn't need both an error return and a special value ?

ldemailly commented on fhttp/http_utils.go:
```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
+		}
+	}
+	log.Debugf("[0. - 100.] for %s roll %f no hit, return default value", input, res)
+	return -1 // default/reminder of probability table
```
this could be used for types other than int (the http status code being an example even if it maps fine to an int), and for cases where -1 is a valid value maybe nil would be a better return for this case
ps: for status the default is ok not -1 and for delay it was 0 but I guess if you document (the function `getWeightedValue` is missing comments/doc even for internal it's useful) clearly maybe it's ok as is (but moves the testing for -1 in the callers)

ldemailly reviewed (commented):
thanks for this work

it's interesting we only save 1/3rd of the 3 copies (-149 lines, +103 lines)

see below

ldemailly committed:
Merge branch 'master' into refactorhttputils

ldemailly committed:
Merge branch 'master' into refactorhttputils

thekuwayama committed:
add WeightValuePair struct to refactor

add lost logging, fix logging comment

ldemailly committed:
Merge branch 'master' into refactorhttputils

ldemailly committed:
Merge branch 'master' into refactorhttputils

ldemailly commented:
friendly ping :)

thekuwayama commented:
Thank you for ping.
Just a second.

thekuwayama committed:
modify WeightValuePair struct; change value type to string

add comments/doc to getWeightedValue function

modify parseFunc => checkConvertible to check that entry is convertible

thekuwayama commented:
Thank you for waiting.

I've tried your suggest, modified `getWeightedValue` to returns string.
`parseFormattedString` should check entry that is convertible before got value from weighted values, so I added `checkConvertible`.

thekuwayama commented:
@ldemailly Sorry reply late.


I think, `parseFromFormattedString` was required to return string if it returned generic value.
`generateStatus` calls strconv.Atoi twice if `parseFromFormattedString` returns string value.
I worry about it.  🤔

thekuwayama commented:
@ldemailly  Thank you very much for reviewing.
It would be difficult for me to refactor so I close this PR.
